### PR TITLE
Renew retention leases while following

### LINKED ISSUE #38718: Prevent CCR followers from falling fatally far behind
A leader shard currently does not provide any guarantees on whether or not it will retain the history that a following shard needs to replicate operations from the leader to the follower. It can happen under normal circumstances and also under error scenarios (network connections are broken, follower goes offline, etc.). In either case though, it is fatal for the follower and the only recourse is file-based recovery which can be expensive. It also means that during this recovery period, the follower is offline, muting its purpose as an extra available copy of the leader shard off-cluster from the leader shard. 

The underlying cause here is the nature of soft deletes. They can be merged away up to a retention limit. Currently the retention limit is based on a fixed number of operations, but this is too difficult for users to reason about (e.g., if I want my leader history to be available in case the follower is offline for up to twelve hours, how many operations should I retain?). Instead, we have to use shard history retention leases for the follower to take a lease on the history of operations on the leader shard and provide guarantees up to some time limit.

Relates #37165

### PR DESCRIPTION
**Pull request by jasontedor**
This commit is the final piece of the integration of CCR with retention leases. Namely, we periodically renew retention leases and advance the retaining sequence number while following.

Relates #37165
Closes #38718

### CODE REVIEW

**Review by dnhatn — APPROVED** [review]
I left a question and two small nits but this looks really good. Thanks @jasontedor.

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java [hunk 1]
```diff
@@ -245,6 +256,88 @@ protected void innerSendShardChangesRequest(long from, int maxOperationCount, Co
...
+                        e -> {
+                            final Throwable cause = ExceptionsHelper.unwrapCause(e);
+                            logRetentionLeaseFailure(retentionLeaseId, cause);
+                            // noinspection StatementWithEmptyBody
+                            if (cause instanceof RetentionLeaseNotFoundException) {
+                                logger.trace(
+                                        "{} background adding retention lease [{}] while following",
+                                        params.getFollowShardId(),
+                                        retentionLeaseId);
+                                CcrRetentionLeases.asyncAddRetentionLease(
```
[the remark this answers is not in the document]
Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java]: I pushed c1351b155532769850703456577070119875ae9b.

[the remark this answers is not in the document]
Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowTasksExecutor.java]: I pushed c1351b155532769850703456577070119875ae9b.

File: x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/CcrRetentionLeaseIT.java [hunk 2]
```diff
@@ -560,6 +526,419 @@ public void testUnfollowFailsToRemoveRetentionLeases() throws Exception {
...
+     * checks if the shard follow node task is cancelled or completed.
+     *
+     * To test this this behavior is correct, we capture the call to renew the retention lease. Then, we will step in between and execute
+     * an unfollow request. This will remove the retention lease on the leader. At this point, we can unlatch the renew call, which will
+     * now be met with a retention lease not found exception. We will cheat and wait for that response to come back, and then synchronously
+     * trigger the listener which will check to see if the shard follow node task is cancelled or completed, and if not, add the retention
+     * lease back. After that listener returns, we can check to see if a retention lease exists on the leader.
+     *
+     * Note, this done mean that listener will fire twice, once in our onResponseReceived hook, and once after our onResponseReceived
+     * callback returns. 🤷‍♀️
```
Comment (jasontedor) [inline x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/CcrRetentionLeaseIT.java]: @tbrooks8 Is there a better way?

**Review by martijnvg — APPROVED** [review]
LGTM - Great to see the last piece of the machinery being added!

File: x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java [hunk 3]
```diff
@@ -151,6 +158,14 @@ void start(
                 coordinateReads();
             });
         });
+
+        synchronized (this) {
+            renewable = scheduleBackgroundRetentionLeaseRenewal(() -> {
```
[the remark this answers is not in the document]
Comment (jasontedor) [reply x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java]: I pushed a commit that does this.

**Pull request merged**